### PR TITLE
feat: add exported Parse(...) which returns a WhereClause

### PR DESCRIPTION
**Pull request by jimlambrt**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,22 +2,57 @@
...
+```
+
+### github.com/hashicorp/go-dbw
+
+```Go
+w, err := mql.Parse("name=alice or name=bob)",User{})
+if err != nil {
+    return nil, err
+}
+err := rw.SearchWhere(ctx, &users, w.Condition,w.Args)
```
Comment (johanbrandhorst) [inline README.md:39]: Nit (styling)
```suggestion
err := rw.SearchWhere(ctx, &users, w.Condition, w.Args)
```

File: README.md [hunk 2]
```diff
@@ -26,14 +61,44 @@ Comparisons can be combined using: `and`, `or`.
...
 contacting us at  security@hashicorp.com.
 
+### Ignoring fields
+
+If your model (Go struct) has fields you don't want users searching then you can
+optionally provide a list of columns to be ignored via `WithIgnoreFields(...)`
+
+### Custom converters/validators
+
+Sometimes the default out of the box bits don't fit your needs.  If you need to
```
Comment (johanbrandhorst) [inline README.md:97]: Nit, helps me with reading this sentence
```suggestion
Sometimes the default out-of-the-box bits don't fit your needs.  If you need to
```

File: README.md [hunk 3]
```diff
@@ -26,14 +61,44 @@ Comparisons can be combined using: `and`, `or`.
...
+override how expressions (column name, operator and value) is converted and
+validated during the generation of a WhereClause, then you can optionally
+provide your own validator/converted via `WithConverter(...)`
```
Comment (johanbrandhorst) [inline README.md:100]:
```suggestion
provide your own validator/converted via `WithConverter(...)`.
```

File: mql.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+// WhereClause contains a SQL where clause condition and its arguments.
+type WhereClause struct {
+	// Condition is the where clause condition
+	Condition string
+	// Args for the where clause condition
+	Args []any
+}
+
+// Parse will parse the query and use the provide database model to create a
```
Comment (johanbrandhorst) [inline mql.go:19]:
```suggestion
// Parse will parse the query and use the provided database model to create a
```

File: mql.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	e, err := exprToWhereClause(expr, fValidators, opt...)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	return e, nil
+}
+
+// exprToWhereClause generates the where clause condition along with it's
```
Comment (johanbrandhorst) [inline mql.go:45]:
```suggestion
// exprToWhereClause generates the where clause condition along with its
```

File: mql.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	case isNil(fValidators):
+		return nil, fmt.Errorf("%s: missing validators: %w", op, ErrInvalidParameter)
+	}
+
+	switch v := e.(type) {
+	case *comparisonExpr:
+		opts, err := getOpts(opt...)
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
```
Comment (johanbrandhorst) [inline mql.go:61]: Doesn't it make more sense to do this before the type switch?

File: mql.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		switch {
+		case opts.withValidateConvertColumn == v.column && !isNil(opts.withValidateConvertFn):
+			return opts.withValidateConvertFn(v.column, v.comparisonOp, v.value)
+		default:
+			columnName := strings.ToLower(v.column)
+			if n, ok := opts.withColumnMap[columnName]; ok {
+				columnName = n
+			}
+			validator, ok := fValidators[strings.ToLower(strcase.ToCamel(columnName))]
```
Comment (johanbrandhorst) [inline mql.go:70]: Why do we need to convert to camel case and then lower case it? Is this some normalization method?

File: mql.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+	"github.com/iancoleman/strcase"
+)
+
+// WhereClause contains a SQL where clause condition and its arguments.
+type WhereClause struct {
+	// Condition is the where clause condition
+	Condition string
+	// Args for the where clause condition
+	Args []any
```
Comment (johanbrandhorst) [inline mql.go:16]: Can we constrain this type?

File: options.go [hunk 9]
```diff
@@ -31,3 +41,52 @@ func withSkipWhitespace() Option {
...
+			return fmt.Errorf("%s: missing field name: %w", op, ErrInvalidParameter)
+		case fieldName != "" && isNil(fn):
+			return fmt.Errorf("%s: missing ConvertToSqlFunc: %w", op, ErrInvalidParameter)
+		}
+		return nil
+	}
+}
+
+// WithIgnoredFields provides an optional list of fields to ignore in the model
+// when parsing. Note: Field names are case sensitive.
```
Comment (johanbrandhorst) [inline options.go:80]: Worth saying whether this is Go field names or SQL column names.

File: README.md [hunk 10]
```diff
@@ -29,6 +29,20 @@ More complex queries can be created using parentheses.
 Example query: 
 `name=alice and age > 11 and (region % Boston or region="south shore")` 
 
+If your model contains a time.Time field, then we'll append `::date` to the
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply README.md:69]: You can specify down to the millisecond: `2023-01-01 14:54:36.896`

File: expr.go [hunk 11]
```diff
@@ -72,19 +74,64 @@ func (e *comparisonExpr) isComplete() bool {
...
+		return nil, fmt.Errorf("%s: %q in %s: %w", op, *e.value, e.String(), ErrInvalidParameter)
+	}
+	if validator.typ == "time" {
+		columnName = fmt.Sprintf("%s::date", columnName)
+	}
+	switch e.comparisonOp {
+	case ContainsOp:
+		return &WhereClause{
+			Condition: fmt.Sprintf("%s like ?", columnName),
+			Args:      []any{fmt.Sprintf("%%%s%%", v)},
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply expr.go:112]: easy to read right?

File: README.md [hunk 12]
```diff
@@ -29,6 +29,20 @@ More complex queries can be created using parentheses.
...
+column name when generating a where clause and the comparison value must be in
+an `ISO-8601` format. Currently, this is the only supported way to compare
+dates, if you need something different then you'll need to provide your own
+custom validator/converter via `WithConverter(...)` when calling
+`mql.Parse(...)`. 
+
+We provide default validation+conversion of for fields in a model when parsing
+and generating a WhereClause. You can provide optional validation+conversion
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply README.md:77]: I meanto to use the `WhereClause` intentionally here, since it's actually what's generated.

File: README.md (hunk 1 above)
Comment (jimlambrt) [reply README.md:39]: ty!

File: expr.go [hunk 13]
```diff
@@ -72,19 +74,64 @@ func (e *comparisonExpr) isComplete() bool {
 	return e.column != "" && e.comparisonOp != "" && e.value != nil
 }
 
+// defaultValidateConvert will validate the comparison expr vale, and then convert the
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply expr.go:77]: ty!

File: mql.go (hunk 4 above)
Comment (jimlambrt) [reply mql.go:19]: ty!

File: mql.go (hunk 8 above)
Comment (jimlambrt) [reply mql.go:16]: I don't think so, especially given that the caller can provide their own
validator/convertor.

File: options.go (hunk 9 above)
Comment (jimlambrt) [reply options.go:80]: Isn't that state with in "fields to ignore in the model"?

File: options.go [hunk 14]
```diff
@@ -31,3 +41,52 @@ func withSkipWhitespace() Option {
...
+		return nil
+	}
+}
+
+// ValidateConvertFunc validates the value and then converts the columnName,
+// comparisonOp and value to a WhereClause
+type ValidateConvertFunc func(columnName string, comparisonOp ComparisonOp, value *string) (*WhereClause, error)
+
+// WithConverter provides an optional ConvertFunc for a column identifier in the
+// query.This allows you to provide whatever custom validation+conversion you
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply options.go:61]: ty!

File: options.go (hunk 9 above)
Comment (johanbrandhorst) [reply options.go:80]: I thought it was a bit ambiguous, I agree fields should really not leave much room for interpretation, but maybe say `struct fields` to make it even more clear?

File: parser_test.go [hunk 15]
```diff
@@ -297,3 +326,20 @@ func Fuzz_parserParse(f *testing.F) {
...
+		err := p.scan(WithConverter("name", nil))
+		require.Error(t, err)
+		assert.ErrorIs(t, err, ErrInvalidParameter)
+		assert.ErrorContains(t, err, "missing ConvertToSqlFunc: invalid parameter")
+	})
+}
+
+func Test_parseComparisonExpr(t *testing.T) {
+	t.Parallel()
+	t.Run("", func(t *testing.T) {})
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply parser_test.go:344]: Ooops. I forgot to delete this.  ty!

File: mql_test.go [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+			want: &mql.WhereClause{
+				Condition: "created_at::date=?",
+				Args:      []any{"2023-01-02"},
+			},
+		},
+		{
+			name:  "success-with-column-map",
+			query: "custom_name=alice",
+			model: testModel{},
+			// opts:  []mql.Option{mql.WithColumnMap(map[string]string{"name": "custom_name"})},
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply mql_test.go:110]: Forgot to delete this. ty!

File: options.go (hunk 9 above)
Comment (jimlambrt) [reply options.go]: better?

File: mql.go (hunk 7 above)
Comment (jimlambrt) [reply mql.go:70]: Example: user query: `member_number = 1` with a model field name: `MemberNumber`
will not match.  Camel case will convert the user provide column name to camel
case to match the "typical" Go field names.  Make sense?

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jimlambrt) [reply README.md]: I meant to to use the `WhereClause` intentionally here, since it's actually what's generated.

File: mql.go (hunk 6 above)
Comment (jimlambrt) [reply mql.go:61]: I don't know.  We only use it this particular case at the moment. We can always
move it in the future if it's needed for other bits in the func.

File: mql.go (hunk 7 above)
Comment (johanbrandhorst) [reply mql.go:70]: Yes that's what I was thinking, so it is a normalization method. If we're just worried about underscores, maybe we could just strip those and avoid the third party dependency?

**Review by johanbrandhorst — APPROVED** [review]

File: mql.go (hunk 7 above)
Comment (jimlambrt) [reply mql.go]: done. ty!

**Pull request merged**